File: TDM.java
public class TDM {
    static int num_documents = 9;
    static int num_terms = 4;
    static int[][] arr = new int[num_documents + 1][num_terms + 1];

    static void create_TDM() {
        for (int i = 0; i < num_documents + 1; i++) {
            arr[i][0] = i;
        }
        for (int j = 0; j < num_terms + 1; j++) {
            arr[0][j] = j;
        }
        for (int i = 1; i < num_documents + 1; i++) {
            for (int j = 1; j < num_terms + 1; j++) {
                arr[i][j] = 0;
            }
        }
    }

    static void updatetermfrequency(int document_index, int term_index, int frequency) {
        arr[document_index][term_index] = frequency;
    }

    static int gettermfrequency(int document_index, int term_index) {
        return arr[document_index][term_index];
    }

    static int getdocumentfrequency(int term_index) {
        int sum = 0;
        for (int i = 1; i < num_documents + 1; i++) {
            if (arr[i][term_index] != 0) {
                sum++;
            }
        }
        return sum;
    }

    static int get_total_terms_in_document(int document_index) {
        int sum = 0;
        for (int j = 1; j < num_terms + 1; j++) {
            sum += arr[document_index][j];
        }
        return sum;
    }

    static void print_TDM() {
        for (int i = 0; i < num_documents + 1; i++) {
            for (int j = 0; j < num_terms + 1; j++) {
                System.out.print(arr[i][j] + "  ");
            }
            System.out.println();
        }
    }

    static void tosparsematrix() {
        int a = 0;
        for (int i = 1; i < num_terms + 1; i++) {
            int b = getdocumentfrequency(i);
            a += b;
            b = 0;
        }
        int[][] sparsematrix = new int[a][3];
        System.out.println("The sparse matrix is");
        System.out.println("D T V");
        for (int i = 1, m = 0; i < num_documents + 1; i++) {
            for (int j = 1; j < num_terms + 1; j++) {
                if (arr[i][j] != 0) {
                    sparsematrix[m][0] = i;
                    sparsematrix[m][1] = j;
                    sparsematrix[m][2] = arr[i][j];
                    m++;
                }
            }
        }
        for (int i = 0; i < a; i++) {
            for (int j = 0; j < 3; j++) {
                System.out.print(sparsematrix[i][j] + " ");
            }
            System.out.println();
        }
    }

    public static void main(String[] args) {
        create_TDM();
        System.out.println("The Term Document matrix is");
        print_TDM();
        System.out.println();
        updatetermfrequency(3, 2, 7);
        updatetermfrequency(3, 1, 8);
        updatetermfrequency(2, 3, 5);
        updatetermfrequency(1, 3, 9);
        updatetermfrequency(3, 3, 9);
        System.out.println("The TDM after updating the term frequencies");
        print_TDM();
        java.util.Scanner scanner = new java.util.Scanner(System.in);
        System.out.print("Enter the term whose document frequency you want to get = ");
        int a = scanner.nextInt();
        System.out.println("The document frequency of " + a + " term is " + getdocumentfrequency(a));
        System.out.print("Enter the number of the document and term whose term frequency you want to get = ");
        int b = scanner.nextInt();
        int c = scanner.nextInt();
        System.out.println("The Doc " + b + " term " + c + ", term frequency is " + gettermfrequency(b, c));
        System.out.print("Enter the document number whose total terms you want to get = ");
        int e = scanner.nextInt();
        System.out.println("The total number of terms in the Doc" + e + " is " + get_total_terms_in_document(e));
        tosparsematrix();
        scanner.close();
    }
}

File: TDM.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int numTerms;
    int numDocuments;
    int **termDocMatrix;
} TermDocumentMatrix;

TermDocumentMatrix createTermDocumentMatrix(int num_terms, int num_documents) {
    TermDocumentMatrix tdm;
    tdm.numTerms = num_terms;
    tdm.numDocuments = num_documents;

    tdm.termDocMatrix = (int **)malloc(num_terms * sizeof(int *));
    for (int i = 0; i < num_terms; i++) {
        tdm.termDocMatrix[i] = (int *)calloc(num_documents, sizeof(int));
    }

    return tdm;
}

void updateTermFrequency(TermDocumentMatrix *tdm, int term_index, int document_index, int frequency) {
    if (term_index >= 0 && term_index < tdm->numTerms && document_index >= 0 && document_index < tdm->numDocuments) {
        tdm->termDocMatrix[term_index][document_index] = frequency;
    }
}

int getTermFrequency(TermDocumentMatrix *tdm, int term_index, int document_index) {
    if (term_index >= 0 && term_index < tdm->numTerms && document_index >= 0 && document_index < tdm->numDocuments) {
        return tdm->termDocMatrix[term_index][document_index];
    } else {
        return 0;
    }
}

int getDocumentFrequency(TermDocumentMatrix *tdm, int term_index) {
    if (term_index >= 0 && term_index < tdm->numTerms) {
        int count = 0;
        for (int i = 0; i < tdm->numDocuments; i++) {
            if (tdm->termDocMatrix[term_index][i] > 0) {
                count++;
            }
        }
        return count;
    } else {
        return 0;
    }
}

int getTotalTermsInDocument(TermDocumentMatrix *tdm, int document_index) {
    if (document_index >= 0 && document_index < tdm->numDocuments) {
        int count = 0;
        for (int i = 0; i < tdm->numTerms; i++) {
            count += tdm->termDocMatrix[i][document_index];
        }
        return count;
    } else {
        return 0;
    }
}

typedef struct {
    int numRows;
    int numCols;
    int **data;
} SparseMatrix;

SparseMatrix toSparseMatrix(TermDocumentMatrix *tdm) {
    SparseMatrix sparseMatrix;
    int a = 0;
    for(int i = 0; i<tdm->numTerms; i++){
        int b = getDocumentFrequency(tdm, i);
        a = a + b;
        b = 0;
    }
    sparseMatrix.numRows = a;
    sparseMatrix.numCols = 3;
    
    sparseMatrix.data = (int **)malloc(a * sizeof(int *));
    for (int i = 0; i < a; i++) {
        sparseMatrix.data[i] = (int *)malloc(3 * sizeof(int));
    }

    printf("The sparse matrix is \n");
    printf("D T V\n");
    for(int i = 0, m = 0; i<tdm->numTerms && m<a; i++){
        for(int j = 0; j<tdm->numDocuments;  j++){
            if(tdm->termDocMatrix[i][j] != 0){   
                sparseMatrix.data[m][0] = j;
                sparseMatrix.data[m][1] = i;
                sparseMatrix.data[m][2] = tdm->termDocMatrix[i][j];
                m++;
            }
        }
    } 
    for(int i = 0; i<a; i++){
        for(int j = 0; j<3; j++){
            printf("%d ", sparseMatrix.data[i][j]);
        }
        printf("\n");
    }
    return sparseMatrix;
}

int print(TermDocumentMatrix *tdm){
     for(int i = 0; i<tdm->numTerms; i++){
        for(int j = 0; j<tdm->numDocuments; j++){
            printf("%d  ", tdm->termDocMatrix[i][j]);
        }
        printf("\n");
    }
    return 0;
}
int main() {
    TermDocumentMatrix tdm = createTermDocumentMatrix(5, 10);
    print(&tdm);
    printf("\n");
    updateTermFrequency(&tdm, 0, 0, 1);
    updateTermFrequency(&tdm, 1, 0, 2);
    updateTermFrequency(&tdm, 2, 1, 3);
    print(&tdm);

    printf("Term Frequency for term 0 in document 0: %d\n", getTermFrequency(&tdm, 0, 0));
    printf("Document Frequency for term 1: %d\n", getDocumentFrequency(&tdm, 1));
    printf("Total Terms in document 0: %d\n", getTotalTermsInDocument(&tdm, 0));

    toSparseMatrix(&tdm);
    printf("Sparse Matrix Representation:\n");

    return 0;
}
